kzg4844/kzg.nim

```diff
@@ -1,21 +1,338 @@
+    return err($res)
+  gCtx.initialized = true
+  return ok()
+
+proc loadTrustedSetup*(fileName: string, precompute: Natural): Result[void, string] =
+  try:
+    let file = open(fileName)
+    result = file.loadTrustedSetup(precompute)
+    file.close()
+  except IOError as ex:
+    return err(ex.msg)
+
+proc loadTrustedSetup*(g1MonomialBytes: openArray[byte],
+                       g1LagrangeBytes: openArray[byte],
+                       g2MonomialBytes: openArray[byte],
+                       precompute: Natural):
+                         Result[void, string] =
+  if gCtx.initialized:
+    return err(TrustedSetupAlreadyLoadedErr)
+  if g1MonomialBytes.len == 0 or g1LagrangeBytes.len == 0 or g2MonomialBytes.len == 0:
+    return err($KZG_BADARGS)
+
+  gCtx.settings = cast[ptr KzgSettings](alloc0(sizeof(KzgSettings)))
+  let res = load_trusted_setup(gCtx.settings,
+      g1MonomialBytes[0].getPtr,
+      g1MonomialBytes.len.uint64,
+      g1LagrangeBytes[0].getPtr,
+      g1LagrangeBytes.len.uint64,
+      g2MonomialBytes[0].getPtr,
+      g2MonomialBytes.len.uint64,
+      precompute.uint64)
+  if res != KZG_OK:
+    dealloc(gCtx.settings)
+    gCtx.settings = nil
+    return err($res)
+  gCtx.initialized = true
+  return ok()
+
+const
+  NumG1 = FIELD_ELEMENTS_PER_BLOB
+  NumG2 = 65
+  G1Len = 48
+  G2Len = 96
+
+type
+  TrustedSetup = object
+    g1MonomialBytes: array[NumG1 * G1Len, byte]
+    g1LagrangeBytes: array[NumG1 * G1Len, byte]
+    g2MonomialBytes: array[NumG2 * G2Len, byte]
+
+proc parseTrustedSetup(input: string): Result[TrustedSetup, string] =
+  var
+    s = newStringStream(input)
+    ts: TrustedSetup
+
+  try:
+    let numG1 = s.readLine().parseInt()
+    if numG1 != NumG1:
+      return err("invalid number of G1 points, expect $1, got $2" % [
+        $NumG1, $numG1
+      ])
+    let numG2 = s.readLine().parseInt()
+    if numG2 != NumG2:
+      return err("invalid number of G2 points, expect $1, got $2" % [
+        $NumG2, $numG2
+      ])
+
+    for i in 0 ..< NumG1:
+      let p = hexToByteArray[G1Len](s.readLine())
+      assign(ts.g1LagrangeBytes.toOpenArray(i * G1Len, ((i + 1) * G1Len) - 1), p)
+
+    for i in 0 ..< NumG2:
+      let p = hexToByteArray[G2Len](s.readLine())
+      assign(ts.g2MonomialBytes.toOpenArray(i * G2Len, ((i + 1) * G2Len) - 1), p)
+
+    for i in 0 ..< NumG1:
+      let p = hexToByteArray[G1Len](s.readLine())
+      assign(ts.g1MonomialBytes.toOpenArray(i * G1Len, ((i + 1) * G1Len) - 1), p)
+
+  except ValueError as ex:
+    return err(ex.msg)
+  except OSError as ex:
+    return err(ex.msg)
+  except IOError as ex:
+    return err(ex.msg)
+
+  ok(ts)
+
+proc loadTrustedSetupFromString*(input: string, precompute: Natural): Result[void, string] =
+  let ts = ?parseTrustedSetup(input)
+  loadTrustedSetup(ts.g1MonomialBytes, ts.g1LagrangeBytes, ts.g2MonomialBytes, precompute)
+
+proc lazyLoadTrustedSetup(): Result[void, string] =
+  when (NimMajor, NimMinor) <= (1,6):
```

Comment:
These two Nim 1.6 issues seem compelling enough to require 2.0 for. It's possible to work around, and @jangko did so, but `const` vs `let` is a real difference, and realistically most of the testing comes from users of this library, which won't be testing with 2.0, never mind 1.6, so creating disparate 1.6-codepaths means creating untested and hard-to-support code.